runtime/src/multi_sig.rs

```diff
@@ -0,0 +1,371 @@
+//! - `change_sigs_required` - Changes the number of signatures required to execute a transaction.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use crate::identity;
+use codec::{Decode, Encode};
+use primitives::{AuthorizationData, AuthorizationError, Key, Signer};
+use rstd::{convert::TryFrom, prelude::*};
+use sr_primitives::{
+    traits::{Dispatchable, Hash},
+    weights::{GetDispatchInfo, Weight},
+    DispatchError,
+};
+use srml_support::{decl_event, decl_module, decl_storage, dispatch::Result, ensure, StorageValue};
+use system::ensure_signed;
+
+pub trait Trait: system::Trait + identity::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Multisig {
+        /// Nonce to ensure unique Multisig addresses are generated. starts from 1.
+        pub MultiSigNonce get(ms_nonce) build(|_| 1u64): u64;
+
+        /// Signers of a multisig. (mulisig, signer) => true/false
+        pub MultiSigSigners get(ms_signers): map (T::AccountId, Signer) => bool;
+        /// Confirmations required before processing a multisig tx
+        pub MultiSigSignsRequired get(ms_signs_required): map T::AccountId => u64;
+        /// Number of transactions proposed in a multisig. Used as tx id. starts from 0
+        pub MultiSigTxDone get(ms_tx_done): map T::AccountId => u64;
+
+        /// Proposals presented for voting to a multisig (multisig, proposal id) => Option<proposal>.
+        /// It is deleted after proposal is processed
+        pub Proposals get(proposals): map (T::AccountId, u64) => Option<T::Proposal>;
+
+        /// Number of votes in favor of a tx. Mapping from (multisig, tx id) => no. of approvals.
+        pub TxApprovals get(tx_approvals): map (T::AccountId, u64) => u64;
+        /// Individual multisig signer votes. (multi sig, signer, )
+        pub Votes get(votes): map (T::AccountId, Signer, u64) => bool;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event() = default;
+
+        /// Creates a multisig
+        ///
+        /// # Arguments
+        /// * `signers` - Signers of the multisig (They need to accept authorization before they are actually added).
+        /// * `sigs_required` - Number of sigs required to process a multi-sig tx.
+        pub fn create_multi_sig(origin, signers: Vec<Signer>, sigs_required: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+            ensure!(signers.len() > 0, "No signers provided");
+            ensure!(u64::try_from(signers.len()).unwrap_or_default() >= sigs_required && sigs_required > 0,
+                "Sigs required out of bounds"
+            );
+            let nonce: u64 = Self::ms_nonce();
+            let new_nonce: u64 = nonce + 1u64;
+            <MultiSigNonce>::put(new_nonce);
+
+            let mut buf = Vec::new();
```

Comment:
We could use `Vec::with_capacity` because we know the maximum size required for this buffer.